llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -21095,6 +21095,50 @@ static SDValue foldTruncStoreOfExt(SelectionDAG &DAG, SDNode *N) {
   return SDValue();
 }
 
+// A custom combine to lower load <3 x i8> as the more efficient sequence
+// below:
+//    ldrb wX, [x0, #2]
+//    ldrh wY, [x0]
+//    orr wX, wY, wX, lsl #16
+//    fmov s0, wX
+//
+static SDValue combineV3I8LoadExt(LoadSDNode *LD, SelectionDAG &DAG) {
+  EVT MemVT = LD->getMemoryVT();
+  if (MemVT != EVT::getVectorVT(*DAG.getContext(), MVT::i8, 3) ||
+      LD->getOriginalAlign() >= 4)
+    return SDValue();
+
+  SDLoc DL(LD);
+  SDValue Chain = LD->getChain();
+  SDValue BasePtr = LD->getBasePtr();
```

Comment:
(Question) May I kindly ask if we should take care of the `offset` of `MachineMemOperand` while we do isel lowering?

Reply:
Unfortunately I am not entirely sure what you mean here? Could. you elaborate?

Reply:
> Unfortunately I am not entirely sure what you mean here? Could. you elaborate?

I'm curious about whether or not should we consider if the `LD` is actually an *indexed* load. Is it possible that it has non UNDEF `offset` SDValue?

Reply:
Ah I see, thanks! I wasn't able to construct a test case where it would be indexed (tried in ff1cde5) but added an assert to catch the use, if possible.

Reply:
> assert to catch the use, if possible.

I think this makes sense because (as per my understanding) usually indexed loads are created while ISelDAGToDAG phase (right after ISelLowering), so I do not expect any assertion fails.